### PR TITLE
Direct Linear Solvers for Symmetric Positive Definite Toeplitz Matrices

### PR DESCRIPTION
**Pull request by SebastianAment**
# Overview 
This PR adds implementations of the Durbin and Levinson algorithms that improve on the performance of StatsBase's implementations by a factor of two for larger matrices. The performance advantage starts to appear before `n = 32` and grow with `n`. Experiments were run on a 2021 MacBook Pro with an M1 Pro.

Since this package specializes on Toeplitz matrices, I think the implementations are better incorporated here than at StatsBase. Further, this PR also adds Trench's algorithm for the efficient inversion of symmetric positive definite Toeplitz matrices. See below for performance benchmarks.

For a reference, see [Golub and Van Loan, page 208.](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwititz36pf1AhXckIkEHWYPAiQQFnoECAUQAQ&url=http%3A%2F%2Fmath.ecnu.edu.cn%2F~jypan%2FTeaching%2Fbooks%2F2013%2520Matrix%2520Computations%25204th.pdf&usg=AOvVaw26GX7osGTu7etg5RSyfh-P)

# Benchmarks

### `n = 4096`
```
durbin
  5.571 ms (2 allocations: 32.05 KiB)
durbin!
  5.568 ms (0 allocations: 0 bytes)
StatsBase.durbin
  11.043 ms (2 allocations: 32.05 KiB)
StatsBase.durbin!
  11.041 ms (0 allocations: 0 bytes)
canonical solve
  161.529 ms (8 allocations: 128.09 MiB)
```

### `n = 32`
```
durbin
  716.241 ns (1 allocation: 336 bytes)
durbin!
  646.341 ns (0 allocations: 0 bytes)
StatsBase.durbin
  818.494 ns (1 allocation: 336 bytes)
StatsBase.durbin!
  794.708 ns (0 allocations: 0 bytes)
canonical solve
  54.375 μs (4 allocations: 9.11 KiB)
```

### `n = 4096`
```
levinson
  9.533 ms (4 allocations: 64.09 KiB)
levinson!
  9.531 ms (2 allocations: 32.05 KiB)
StatsBase.levinson
  23.107 ms (4 allocations: 64.09 KiB)
StatsBase.levinson!
  23.110 ms (0 allocations: 0 bytes)
canonical solve
  161.600 ms (6 allocations: 128.06 MiB)
```

### `n = 32`
```
levinson
  905.273 ns (2 allocations: 672 bytes)
levinson!
  897.925 ns (1 allocation: 336 bytes)
StatsBase.levinson
  1.221 μs (2 allocations: 672 bytes)
StatsBase.levinson!
  1.175 μs (0 allocations: 0 bytes)
direct solve
  52.750 μs (3 allocations: 8.78 KiB)
```

### `n = 4096`
```
trench
  18.469 ms (14 allocations: 128.06 MiB)
canonical inv
  499.677 ms (6 allocations: 130.03 MiB)
```
### `n = 32`
```
trench
  1.279 μs (7 allocations: 8.94 KiB)
canonical inv
  59.542 μs (4 allocations: 24.58 KiB)
```

### CODE REVIEW

**Comment by SebastianAment** [comment]
Further notes:

- `StatsBase.levinson!` mutates `b`, which is unexpected (and should probably be called `levinson!!`).  For this reason, the new `levinson!` implementation still allocates a single vector of length `n`.
-  If the current maintainers want to incorporate the new implementations, I'd also change lines 687 to 701 to use the new `levinson` implementation.
- We could also add a `SymmetricPositiveDefiniteToeplitz` type, for which `inv` and `\` automatically call `trench` and `levinson`, respectively.

File: src/directLinearSolvers.jl
Comment (dlfivefifty) [inline src/directLinearSolvers.jl:19]: Does the `1` cause a type instability?

File: src/directLinearSolvers.jl
Comment (dlfivefifty) [inline src/directLinearSolvers.jl:53]: `!isone(r_0)` may be more general

File: src/directLinearSolvers.jl
Comment (dlfivefifty) [inline src/directLinearSolvers.jl:57]: ditto

File: src/directLinearSolvers.jl
Comment (dlfivefifty) [inline src/directLinearSolvers.jl:110]: ditto

File: src/directLinearSolvers.jl
Comment (dlfivefifty) [inline src/directLinearSolvers.jl:129]: ditto

File: src/directLinearSolvers.jl
Comment (dlfivefifty) [inline src/directLinearSolvers.jl:133]: ditto

File: test/runtests.jl
Comment (dlfivefifty) [inline test/runtests.jl:139]: add tests for type stability (`@inferred`).

**Review by dlfivefifty — CHANGES_REQUESTED** [review]

**Comment by dlfivefifty** [comment]
> * We could also add a `SymmetricPositiveDefiniteToeplitz`

Not a good idea: there’s no equivalent in LinearAlgebra. The analogue there is explicitly calling `  cholesky  `

**Comment by SebastianAment** [comment]
> > * We could also add a `SymmetricPositiveDefiniteToeplitz`
> 
> Not a good idea: there’s no equivalent in LinearAlgebra. The analogue there is explicitly calling ` cholesky`

This is not central to this PR, just adding my thoughts: 

- There are the `Symmetric` and `Hermitian` wrappers in LinearAlgebra which do impact how `factorize` behaves. There's no clear reason in my mind why one property should be dispatched on, while the other one shouldn't. In the case of Toeplitz matrices, checking definiteness is also an O(n^2) operation, same as checking symmetry for regular matrices.
- Signaling positive definiteness via types would make it easier to write generic code that should work efficiently for both definite and indefinite systems. 
- It would be easier for non-experts (i.e. people who have not heard of Trench's or Levinson's algorithms) to take advantage of the efficient solver and inversion algorithms by just turning on a flag. This could be done by having a `PositiveDefinite` trait as a field in the `SymmetricToeplitz` structure.

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:4]:
```suggestion
    durbin(r::AbstractVector)

Computes `y = T \\ (-r)` where `T = SymmetricToeplitz(vcat(1, r[1:end-1]))`.
```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:12]:
```suggestion

    durbin!(y::AbstractVector, r::AbstractVector)

```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:33]:
```suggestion

    trench(T::SymmetricToeplitz)

```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:38]:
```suggestion
    n = length(T.vc)
```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:74]:
```suggestion
    n = LinearAlgebra.checksquare(B)
```

When resolving the merge conflict, I might have removed the import of `checksquare`...

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:93]:
```suggestion

    levinson(r::AbstractVector, b::AbstractVector)

solves `x = T \\ b` where `T = SymmetricToeplitz(vcat(1, r))`, i.e. `T_{ij} = r[abs(i-j)]`
where by assumption `r[0] = 1` and `T` is positive definite.
```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:101]:
```suggestion

    levinson!(x::AbstractVector, r::AbstractVector, b::AbstractVector, y::AbstractVector = zero(x))

Same as `levinson`, but uses the pre-allocated vector `x` to store the result.
```

**Review by dkarrasch — COMMENTED** [review]
I found a few minor style issues.

**Comment by SebastianAment** [comment]
> I found a few minor style issues.

I just committed your suggestions.

**Comment by ParadaCarleton** [comment]
Is anything holding this up? @dkarrasch @SebastianAment

**Review by dlfivefifty — APPROVED** [review]

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:11]:
```suggestion

```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:32]:
```suggestion

```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:47]:
```suggestion
    trench!(B::AbstractMatrix, T::SymmetricToeplitz)

```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:49]:
```suggestion
of `T` and returns a `Symmetric` wrapper of `B`.
```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:94]:
```suggestion
    levinson(r::AbstractVector, b::AbstractVector)

Solves `x = T \\ b` where `T = SymmetricToeplitz(vcat(1, r))`, i.e. `T_{ij} = r[abs(i-j)]`
```

File: src/directLinearSolvers.jl
Comment (dkarrasch) [inline src/directLinearSolvers.jl:100]:
```suggestion

```

**Review by dkarrasch — COMMENTED** [review]
A few more style suggestions.

**Comment by dkarrasch** [comment]
Out of curiosity, why not make the overloaded functions available in `StatsBase.jl`? Do you need special fields of structs that are owned by this package here?

**Comment by ParadaCarleton** [comment]
@dkarrasch if the only thing missing is a couple style changes, perhaps we should just apply the changes and merge, to avoid holding this up too long?

**Comment by dkarrasch** [comment]
> Out of curiosity, why not make the overloaded functions available in `StatsBase.jl`? Do you need special fields of structs that are owned by this package here?

I still wonder... There will be `durbin` methods and others both in `StatsBase.jl` and here. How will people judge which one to use? The backend appears to be fairly generic.

**Comment by ParadaCarleton** [comment]
> > Out of curiosity, why not make the overloaded functions available in `StatsBase.jl`? Do you need special fields of structs that are owned by this package here?
> 
> I still wonder... There will be `durbin` methods and others both in `StatsBase.jl` and here. How will people judge which one to use? The backend appears to be fairly generic.

I do think the best approach is to dispatch on the most efficient function for a given matrix, as @SebastianAment suggested.

**Comment by ParadaCarleton** [comment]
> > Out of curiosity, why not make the overloaded functions available in `StatsBase.jl`? Do you need special fields of structs that are owned by this package here?
> 
> I still wonder... There will be `durbin` methods and others both in `StatsBase.jl` and here. How will people judge which one to use? The backend appears to be fairly generic.

Presumably they will be, but that would require a PR in StatsBase.jl.

**Comment by ParadaCarleton** [comment]
@dlfivefifty could you merge this?

**Pull request merged**
